remove `nonlocal_` modality

riaqn opened the pull request:
As subject. This feature is undocumented and rarely used, and it causes complications for new features (in particular, closure over explicit borrowing).

lpw25 commented on ocaml/typing/typecore.ml:
```diff
@@ -4595,7 +4585,6 @@ and type_expect_
           None -> None
         | Some sexp ->
             if !Clflags.principal then begin_def ();
-            (* TODO: mode can be more relaxed than this if fields are nonlocal *)
```
I think this comment still holds if we replace `nonlocal` with `global`

lpw25 commented on ocaml/typing/typedecl.ml:
```diff
@@ -368,12 +368,9 @@ let transl_global_flags loc attrs =
   let global = transl_global_flag loc (Builtin_attributes.has_global attrs) in
-  let nonlocal = transl_global_flag loc (Builtin_attributes.has_nonlocal attrs) in
-  match global, nonlocal with
-  | true, true -> raise(Error(loc, Global_and_nonlocal))
```
The `Global_and_nonlocal` error constructor also needs to be removed.

lpw25 commented on ocaml/typing/includecore.ml:
```diff
@@ -492,16 +492,11 @@ let report_type_mismatch first second decl env ppf err =
   match flag0, flag1 with
-  | Global, (Nonlocal | Unrestricted) ->
+  | Global, Unrestricted ->
     Some {order = First; nonlocal = false}
```
The `nonlocal` field should be removed. (Should also generally grep the source for "nonlocal" to see if there are any other leftovers.)

lpw25 reviewed (approved):
This looks good, although there are still some other remnants around that should be removed too.

riaqn commented:
Thanks - I have addresses the comments. The remainining mentioning of `nonlocal` are (I believe) those in middle and backend, those meaning "not local", which I believe is unrelated to the nonlocal modality.
